{% extends "layout.html" %}
{% block body %}
<div id="contents">
  <h2 id="usage">Basic Usage</h2>
  <div>
    <p>
    PatScanUI provides a graphical interface for the PatScan tool. See the <a
      href="{{ url_for('about') }}">about page</a> for the relevant references.
    </p>
    <p>
    As a first step, you need to upload a sequence file in FASTA format.
    This is the sequence you can search using PatScan. The file will be
    kept on the server as long as you work on it and deleted an hour
    after you finished using the software.
    </p>
    <p>
    Once your file is uploaded, you can specify a pattern by dragging
    pattern elements from the menu bar on the right side of the screen
    to the work area. Once the pattern element is on the work area, you
    can fill in details, and reorder the elements by dragging them
    around.
    </p>
    <p>
    Above the pattern area, there is a configuration menu containing a
    couple of buttons to adjust the software.
    </p>

    <p>
    The <b>DNA/Protein</b> selector allows you to tell PatScan what
    kind of sequence file you uploaded.  This affects the selection of
    available pattern elements.
    </p>

    <p>
    While working on the pattern, you can switch on the <b>continuous
      preview</b>, that will regularly search your uploaded sequence
    for the currently selected pattern. This will allow you to tweak
    the pattern until you get the desired result.
    </br>

    <p>
    If you want to get rid of the current pattern elements, you can
    clear all of them using the <b>clear</b> button. Alternatively, you
    may drag single elements onto the trash on the right hand menu.
    </p>

    <p>
    If you want to get an example of many of the possible pattern elements,
    you can load an <b>example</b>.
    </p>

    <p>
    Below the work area, you can expand the <b>computed pattern</b>
    view, in case you need to tweak the generated pattern manually
    </p>
  </div>

  <h2 id="patterns">Available Patterns</h2>
  <p>
    PatScan supports many different pattern types, here we will explain how they work.
  </p>

  <div id="accordion">
    <h3 id="string"><a href="#">String Pattern</a></h3>
    <div>
      <p>
        String patterns allow you to specify a DNA/RNA/protein sequence to search for.
        This is the most basic pattern to use. For DNA sequences, in addition to
        the regular <code>ACGT</code> nucleotide letters, you can also use the
        <a class="reference" href="#codes">IUPAC ambiguity codes</a>.
        String patterns can be matched while allowing <a class="reference" href="#variations">variations</a>.
      </p>
    </div>

    <h3 id="range"><a href="#">Range Pattern (insertion range)</a></h3>
    <div>
      <p>
      Range patterns are a shorthand for using a <a class="reference"
        href="#string">string pattern</a> consisting of a run of
      arbitrary nucleotides or amino acids. They are especially useful to allow
      for a variable size of insertions. So if you e.g. wanted to allow for <strong>20</strong>
      to <strong>30</strong> arbitrary nucleotides, you can use a range pattern from
      <strong>20</strong> to <strong>30</strong> instead of using a string pattern of 20
      <code>N</code>s and allowing up to 10 insertions.
      </p>
    </div>

    <h3 id="complement"><a href="#">Complement Pattern</a></h3>
    <div>
      <p>
      Complement patterns match the reverse complement of a
      <a class="reference" href="#named">named</a> pattern. Thus, they only
      apply for DNA and RNA sequences. It is possible to specify
      <a class="reference" href="#altcomplementrule">alternative complementation rules</a>
      to specify which bases can be complements. Complements can also have some
      <a class="referece" href="#variations">variations</a>.
      </p>
    </div>

    <h3 id="repeat"><a href="#">Repeat Pattern</a></h3>
    <div>
      <p>
      Repeat patters repeat a previously specified
      <a class="reference" href="#named">named</a> pattern. It can have
      <a class="reference" href="#variations">variations</a> compared to the
      original pattern.
      </p>
    </div>

    <h3 id="alternative"><a href="#">Alternative Pattern</a></h3>
    <div>
      <p>
      An alternative pattern contains any two other patterns and requires that one
      or the other (or potentially both) get a match.
      </p>
    </div>

    <h3 id="lengthlimit"><a href="#">Length limit Pattern</a></h3>
    <div>
      <p>
      The length limit pattern allows to specify that the length of
      any <a class="reference" href="#named">named</a> pattern(s) is/are less
      than a specified number of bases or amino acids. Use shift-click to select
      multiple named patterns from the list.
      </p>
    </div>

    <h3 id="weight"><a href="#">Weight Pattern</a></h3>
    <div>
      <p>
      The weight pattern allows users to specify a position specific probability/weight matrix
      for DNA sequences. For every position in the matrix, the probabilities of the individual bases (in %)
      should add up to 100. The final weight of the pattern is calculated from the percent values of the matching bases.
      </p>
    </div>

    <h3 id="anyof"><a href="#">Any-Of Pattern</a></h3>
    <div>
      <p>
      The Any-Of pattern allows matching a list of amino acids for a given position.
      </p>
    </div>

    <h3 id="notanyof"><a href="#">Not-Any-Of Pattern</a></h3>
    <div>
      <p>
      The Not-Any-Of pattern is the inverse of the <a class="reference" href="#anyof">Any-Of pattern</a>
      and allows specifying a list of amino acids that can not occur at a given position.
      </p>
    </div>

    <h3 id="altcomplementrule"><a href="#">Alternative Complementation Rule</a></h3>
    <div>
      <p>
      Alternative complementation rules allow creating base pairings that are not covered by the
      conventional <code>AT</code>, <code>TA</code>, <code>CG</code>, or <code>GC</code> pairings.
      If you want to allow for e.g. <code>GU</code> and <code>UG</code> pairings in RNA, you can create
      an alternative complementation rule for all of these pairings.
      </p>
    </div>

    <h3 id="variations"><a href="#">Variations</a></h3>
    <div>
      <p>
        When patterns allow for variations, you can specify the number of mutations (mismatches),
        insertions or deletions you want to allow.
      </p>
    </div>

    <h3 id="named"><a href="#">Names</a></h3>
    <div>
      <p>
      Patterns can be assigend names by clicking the <em>unnamed</em> button.
      Named patterns can be used in <a class="reference" href="#complement">complement patterns</a> and
      <a class="reference" href="#repeat">repeat patterns</a>.
      </p>
    </div>


    <h3 id="codes"><a href="#">Nucleotide Ambiguity Codes</a></h3>
    <div>
      <p>
      <table id="ambiguity-table">
        <thead>
          <tr>
            <td class="legend">Code</td>
            <td>M</td>
            <td>R</td>
            <td>W</td>
            <td>S</td>
            <td>Y</td>
            <td>K</td>
            <td>V</td>
            <td>H</td>
            <td>D</td>
            <td>B</td>
            <td>N</td>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td class="legend">Nucleotides</td>
            <td>AC</td>
            <td>AG</td>
            <td>AT</td>
            <td>CG</td>
            <td>CT</td>
            <td>GT</td>
            <td>ACG</td>
            <td>ACT</td>
            <td>AGT</td>
            <td>CGT</td>
            <td>ACGT</td>
          </tr>
        </tbody>
      </table>
      </p>
    </div>
  </div>

</div>
{% endblock %}
{% block scripts %}
<script src="{{ url_for('static', filename="js/jquery-ui-1.8.22.custom.min.js") }}"></script>
<script src="{{ url_for('static', filename="js/jquery.ui.touch-punch.min.js") }}"></script>
<script type="text/javascript">

function findTopicIdx(topic) {
  var topics = $('#accordion').children('h3');
  for (var i=0; i < topics.length; i++) {
    if (topics[i].id == topic) {
      return i;
    };
  }
  return -1;
}

function setActiveTopic(topic) {
  var idx = findTopicIdx(topic);
  if (idx < 0) return;
  $('#accordion').accordion('option', 'active', idx);
}

function hashChangeCallback() {
  setActiveTopic(window.location.hash.slice(1));
}

$(document).ready(function() {
  $('#accordion').accordion({
    active: false
  });
  window.addEventListener("hashchange", hashChangeCallback, false);
  setActiveTopic(window.location.hash.slice(1));
});
</script>
{% endblock %}
