{% extends "layout.html" %}
{% block body %}
<div id="contents">

  This tutorial will explain the use of <a href="{{ url_for('about') }}">PatScan</a> patterns using biologically
  relevant examples. If you want to follow along, download the input files for
  <a href="{{ url_for('static', filename='tutorial/example1_input.fa') }}">example 1</a> and
  <a href="{{ url_for('static', filename='tutorial/example2_input.fa') }}">example 2</a>.

  <div id="pattern-selections">
    <ol id="tutorial-nav">
      <li>Identifying Iron-Response Elements in mRNA
        <ol>
          <li><a href="#example1-1">String and Complement Patterns</a></li>
          <li><a href="#example1-2">Variations and Range Patterns</a></li>
          <li><a href="#example1-3">Alternative Patterns</a></li>
          <li><a href="#example1-4">Alternative Complementation Rules</a></li>
          <li><a href="#example1-5">Weight Patterns</a></li>
        </ol>
      </li>
      <li>Indentifying Cytochrome P450 Cysteine Heme-Iron Ligand Signatures
        <ol>
          <li><a href="#example2-1">Any-Of and Not-Any-Of Patterns</a></li>
        </ol>
    </ol>
  </div>

  <h2 id="example1">Example 1: Identifying Iron-Response Elements in mRNA.</h2>
  <div>
    <p>
    Besides its catalytic function in the TCA cycle, the Aconitase enzyme has an
    important regulatory function in both Eukaryotes and Prokaryotes. When
    active as a regulator, Aconitase binds specific structural motifs in mRNA,
    so-called Iron-Response Elements (IREs). In Eukaryotes, it is highly
    sensitive to iron deprivation and oxidative stress. In bacteria, the
    role of Aconitase seems to be similar, though the recognition sequence
    is less preserved. Aconitase recognises the RNA stem loop structure of
    the IRE consisting of a C bulge, a stem of 6 amino acids, and a loop of 5 or
    6 nucleotides. In Eukaryotes, the loop sequence always is CAGUG, in
    Bacteria, more variations are possible.<br>
    In this example, we will build a set of patterns to identify putative IREs
    in the 5’ UTR sequences of the bacterium Streptomyces coelicolor. First
    starting with a relatively basic example, we will slowly extend it to cover
    more advanced patterns and use cases.
    </p>
    <h3 class="example" id="example1-1">1. String and Complement Patterns</h3>
    <p>
    The basic IRE pattern is
    <code>CN<sub>1</sub>N<sub>2</sub>N<sub>3</sub>N<sub>4</sub>N<sub>5</sub>N<sub>6</sub>CAGUGN'<sub>6</sub>N'<sub>5</sub>N'<sub>4</sub>N'<sub>3</sub>N'<sub>2</sub>N'<sub>1</sub></code>,
    where <code>N<sub>x</sub></code> is any amino acid, and <code>N'<sub>x</sub></code> is its complement.<br>
    An initial attempt at finding such a pattern using PatScan would use a number of <a href="{{ url_for('help') }}#string">string patterns</a> and a
    <a href="{{ url_for('help') }}#complement">complement pattern</a>. Just using a string pattern with <code>CNNNNNNCAGUGNNNNNN</code> would fail to ensure the second
    set of <code>N</code>s would be the reverse complement of the first, so instead we need to break the pattern up a bit.
    </p>

    <div class="tutorial-box">
      <div class="tutorial-content">
        On the start page , hit the "Browse" button to upload the
        <code>example1_input.fa</code> file.
      </div>
      <div class="figure">
        <img src="{{ url_for('static', filename='tutorial/start_page.png') }}"
             alt="start page">
        <div class="figure-text">Fig 1: Start page.</div>
      </div>
    </div>

    <div class="tutorial-box">
      <div class="tutorial-content">
        You will end up on an empty pattern input screen.
      </div>
      <div class="figure">
        <img src="{{ url_for('static', filename='tutorial/empty.png') }}"
             alt="Empty pattern field" width="550px">
        <div class="figure-text">Fig 2: Empty pattern input screen.</div>
      </div>
    </div>

    <div class="tutorial-box">
      <div class="tutorial-content">
        In order to add patterns, we use the pattern selector, as shown in Fig.
        3. Patterns can be added to the work area by dragging them from the
        selector, or simply by clicking on them. If you need to rearrange
        patterns, you can simply move them around using drag and drop in the
        work area.
      </div>
      <div class="figure">
        <img src="{{ url_for('static', filename='tutorial/pattern_selector.png') }}"
             alt="Pattern selector" width="550px">
        <div class="figure-text">Fig 3: Pattern selector.</div>
      </div>
    </div>

    <div class="tutorial-box">
      <div class="tutorial-content">
        Select three string patterns and one complement patterns. For the first
        string pattern, select <code>C</code>. For the second string pattern,
        select <code>NNNNNN</code>. For the third string pattern, select
        <code>CAGUG</code>. Click the "unnamed" button next to the second string
        pattern. The button text will change to "p1", the identifier of this
        pattern. In the complement pattern, you can now select the “p1”
        identifier for the “complement of” dropdown. Once you are done, your
        work area should look like in Fig. 4.
      </div>
      <div class="figure">
        <img src="{{ url_for('static', filename='tutorial/example1_1.png') }}"
             alt="Using string and complement patterns" width="550px">
        <div class="figure-text">Fig 4: Using string and complement patterns.</div>
      </div>
    </div>
    <div class="clearfix"></div>

    <h3 class="example" id="example1-2">2. Variations and Range patterns</h3>

    <div class="tutorial-box">
      <div class="tutorial-content">
      If we run the pattern we created in Example 1.1, we unfortunately don't
      get any hits in the input data. The reason for this is that the IRE motif
      in Bacteria is not as conserved. Notably, the stem region does not need to
      have six perfect matches. To allow for a mismatch in the stem, hit the
      "Variations" button on the complement pattern. A number of new fields
      appear allowing us to specify how many mismatches, insertions or deletions
      this pattern can tolerate. Put a "1" in the "mismatches" field and ignore
      the other fields to just allow a single mismatch.
      </div>
      <div class="figure">
        <img src="{{ url_for('static', filename='tutorial/variations.png') }}"
             alt="Allowing one mutation" width="550px">
        <div class="figure-text">Fig 5: Allowing one mutation.</div>
      </div>
    </div>

    <div class="tutorial-box">
      <div class="tutorial-content">
        There is one more detail about the IRE pattern that we don't cover yet.
        The loop region acually can contain an optional sixth base after the
        <code>CAGUG</code> pattern. Let's add support for this by adding a <a
           href="{{ url_for('help') }}#range">range pattern</a>. Drag and drop
         the range pattern field to sit between the string pattern for the loop
         region and the repeat pattern for the reverse part of the stem loop.
         Set the range pattern to cover from "0" to "1" bases.
      </div>
      <div class="figure">
        <img src="{{ url_for('static', filename='tutorial/range_pattern.png') }}"
             alt="The loop range pattern" width="550px">
        <div class="figure-text">Fig 6: The loop range pattern in place.</div>
      </div>
    </div>

    <div class="tutorial-box">
      <div class="tutorial-content">
        Another variation that is common on bacterial IREs is that the stem of
        the stem loop structure is only five bases long, not six. We can use
        another range pattern to support this. Hit the "Remove" button on the
        <code>NNNNNN</code> string pattern, and drag another range pattern into
        the same location instead. Remember to hit the "unnamed" button again so
        it says "p1" as before. You also need to select that from the
        "Complement of" dropdown of the complement pattern again.
      </div>
      <div class="figure">
        <img src="{{ url_for('static', filename='tutorial/example1_2.png') }}"
             alt="Replacing the N string pattern with a range pattern" width="550px">
        <div class="figure-text">Fig 7: Replacing the N string pattern with a
          range pattern.</div>
      </div>
    </div>

    <div class="clearfix"></div>

    <h3 class="example" id="example1-3">3. Alternative Patterns</h3>

    <div class="tutorial-box">
      <div class="tutorial-content">
        Not only the stem region but also the loop region can have variations in
        Bacteria. Notably, the loop sequence <code>GAGAG</code> has also been
        reported to be functional. To support searching for <code>CAGUG</code>
        or <code>GAGAG</code> at the same time, we can use the
        <a href="{{ url_for('help') }}#alternative">alternative pattern</a>.
        It is basically a container that can hold any two other patterns, one of
        which has to match. To add it, drag it into position next to the
        <code>CAGUG</code> string pattern, and then drag the <code>CAGUG</code>
        pattern into the alternative pattern. Drag a new string pattern into the
        remaining free slot of the alternative pattern, and set that sequence to
        <code>GAGAG</code>. It does not matter what order the two string patterns
        have inside the alternative pattern.
      </div>
      <div class="figure">
        <img src="{{ url_for('static', filename='tutorial/example1_3.png') }}"
             alt="Using the alternative pattern" width="550px">
        <div class="figure-text">Fig 8: Using the alternative pattern.</div>
      </div>
    </div>

    <div class="clearfix"></div>

    <h3 class="example" id="example1-4">4. Alternative Complementation Rules</h3>

    <div class="tutorial-box">
      <div class="tutorial-content">
        By default, PatScan uses the regular DNA complementation rules to
        determine which bases can be in a reverse complement. IREs are RNA
        sequences, so additional <code>GU/UG</code> pairs are possible.
        To support this, you can create custom
        <a href="{{ url_for('help') }}#altcomplementrule">alternative
          complementation rules</a>. Note that due to constraints in PatScan,
        the alternative complementation rules need to be specified before being
        usable in a <a href="{{ url_for('help') }}#complement">complement
          rule</a>. Personally, I prefer having all alternative complementation
        rules at the start of my pattern set.<br>
        To add a complement pair, hit the "Add Complement Pair" button. You need
        to add a pair for every "direction" you want to allow complementation
        in. That is, adding e.g. a <code>GU</code> pair does not set up pairing
        rules for <code>UG</code>.
      </div>
      <div class="figure">
        <img src="{{ url_for('static', filename='tutorial/example1_4.png') }}"
             alt="Using alternative complement rules" width="550px">
        <div class="figure-text">Fig 9: Using alternative complement rules.</div>
      </div>
    </div>

    <div class="clearfix"></div>

    <h3 class="example" id="example1-5">5. Weight Patterns</h3>

    <div class="tutorial-box">
      <div class="tutorial-content">
        If we wanted to explore further alternatives to the loop pattern
        sequence, we could put alternative patterns into alternative patterns to
        create a deeper logical branching. That approach does get a bit
        unwieldy, though. A better option is to instead use a
        <a href="{{ url_for('help') }}#weight">weight pattern</a>. A weight
        pattern is a kind of position-specific probability matrix that includes
        a cut-off value.<br>
        Weights can be added by clicking the "Add Weight" button. Every weight
        expects the probability values in percent for encountering a specific
        base at a given position. If the probability is zero, no value needs to
        be entered. The final weight is the summed-up probabilities of the whole
        sequence.<br>
        If we wanted to build a weight pattern to capture the loop sequences
        <code>CAGUG</code> and <code>GAGAG</code> but also want to allow for a
        <code>G</code> base in the fourth position, we would use the following
        approach:<br>
        First, remove the old alternative pattern by clicking its "Remove"
        button. Then, drag and drop a weight pattern in that place instead.
        Hit the "Add weight" button five times to add weights for all positions
        of our desired loop sequence. We want the first base to be
        <code>C</code> or <code>G</code>, so in the first column, we fill in
        "50" in the second and third rows. Zeroes can be left out, so no need to
        fill out the other rows. The second base should always be an
        <code>A</code>, so we put "100" into the first row of the second column
        and leave the rest blank. The third base should always be a
        <code>G</code>, so "100" goes into the third row of the third column.
        The fourth base could be <code>A</code>, <code>G</code>, or
        <code>T</code> (actually <code>U</code>, but for weight matrices those
        bases are equivalent). The weights are a bit arbitrary for this one, as
        long as <code>C</code> gets a zero. As there is no published evidence
        for <code>G</code> actualy being valid, I personally like to score this
        a bit less than the experimentally validated options. So, for column
        four, let's put "40" into the first and last rows, and "20" into the
        third row. For the last base, we again always want a <code>G</code>, so
        the fifth column gets a "100" in the third row.<br>
        To calculate the weight cut-off, we use the minimal score a valid
        pattern can obtain. There are three bases at "100", one base at "50" and
        one where the lowest valid score is "20", so "370" is the minimum score
        we expect a good pattern to hit.
      </div>
      <div class="figure">
        <img src="{{ url_for('static', filename='tutorial/example1_5.png') }}"
             alt="Using a weight pattern" width="550px">
        <div class="figure-text">Fig 10: Using a weight pattern.</div>
      </div>
    </div>

    <div class="clearfix"></div>

  </div>

  <h2>Example 2: Identifying Cytochrome P450 Cysteine Heme-Iron Ligand Signatures (PS00086)</h2>
  <div>
    Cytochrome P450 proteins utilise heme cofactors to perform oxidation reactions
    and are found all over the tree of life. The heme cofactor binds to a cysteine
    in the active site of the enzyme.
    In this example, we will build a set of patterns to identify the active site
    cysteine of cytochrome P450 enzymes found in the proteome of the bacterium
    <i>Streptomyces coelicolor</i>. We assume that you have already gone over
    example 1 and are familiar with the basic pattern creation.

    <h3>1. Any-Of and Not-Any-Of Patterns</h3>
    <p>
    According to the <a href="https://prosite.expasy.org/PS00086">PROSITE database entry PS00086</a>,
    the pattern for identifying the active site cysteine is
    <code>[FW]-[SGNH]-x-[GD]-{F}-[RKHPT]-{P}-C-[LIVMFAP]-[GAD]</code>. If you
    are not familiar with the PROSITE pattern notation, amino acid letters
    wrapped in square brackets (<code>[ ]</code>) are alternatives, amino
    acid letters wrapped by curly braces (<code>{ }</code>) should not be present
    at a given position, and <code>x</code> means any arbirary amino acid.
    </p>
  </div>
</div>
<div id="dialog" title="An Image!" style="display: none;">
  <img id="image" src=""/>
</div>
{% endblock %}
{% block scripts %}

<script src="{{ url_for('static', filename='js/jquery-ui-1.8.22.custom.min.js') }}"></script>
<script src="{{ url_for('static', filename='js/jquery.ui.touch-punch.min.js') }}"></script>
<script type="text/javascript">

function imagePopup(title, uri) {

  var image_dialog = $("#dialog");
  var image_tag = $('#image');

  //Set the image src
  image_tag.attr('src', uri);

  //When the image has loaded, display the dialog
  image_tag.load(function(){

    $('#dialog').dialog({
      modal: true,
      resizable: false,
      draggable: false,
      width: 'auto',
      title: title
    });
  });
};

function activateImagePopup(e) {
  e.preventDefault();
  imagePopup($(this).attr('alt'), $(this).attr('src'));
};

$(document).ready(function() {
  $('.figure').children('img').click(activateImagePopup);
});
</script>
{% endblock %}
